packages/core/src/indexing-store/historical.ts

```diff
@@ -1,363 +1,893 @@
+import { type Hex, padHex } from "viem";
+import { getReadonlyStore } from "./readonly.js";
+import type { HistoricalStore, OrderByInput, WhereInput } from "./store.js";
+import {
+  decodeRecord,
+  encodeField,
+  encodeRecord,
+  validateRecord,
+} from "./utils/encoding.js";
 import { parseStoreError } from "./utils/errors.js";
 import { buildWhereConditions } from "./utils/filter.js";
 
-const MAX_BATCH_SIZE = 1_000 as const;
+const MAX_BATCH_SIZE = 1_000;
+const CACHE_FLUSH = 0.35;
+
+type Entry<record = UserRecord> = {
+  opIndex: number;
+  bytes: number;
+  record: record;
+};
+
+type StoreCache = {
+  [tableName: string]: {
+    insert: { [id: string | number]: Entry };
+    update: { [id: string | number]: Entry };
+    find: { [id: string | number]: Entry<UserRecord | undefined> };
+  };
+};
 
 export const getHistoricalStore = ({
-  kind,
+  encoding,
   schema,
   namespaceInfo,
   db,
+  common,
+  isCacheFull: _isCacheFull,
 }: {
-  kind: "sqlite" | "postgres";
+  encoding: "sqlite" | "postgres";
   schema: Schema;
   namespaceInfo: NamespaceInfo;
   db: HeadlessKysely<any>;
-}): WriteStore<"historical"> => ({
-  create: async ({
-    tableName,
-    id,
-    data = {},
-  }: {
-    tableName: string;
-    id: UserId;
-    data?: Omit<UserRecord, "id">;
-  }) => {
-    const table = (schema[tableName] as { table: Table }).table;
-
-    return db.wrap({ method: `${tableName}.create` }, async () => {
-      const createRow = encodeRow({ id, ...data }, table, kind);
-
-      const row = await db
-        .withSchema(namespaceInfo.userNamespace)
-        .insertInto(tableName)
-        .values(createRow)
-        .returningAll()
-        .executeTakeFirstOrThrow()
-        .catch((err) => {
-          throw parseStoreError(err, { id, ...data });
-        });
+  common: Common;
+  isCacheFull: boolean;
+}): HistoricalStore => {
+  const maxSizeBytes = common.options.indexingCacheBytes;
+  const storeCache: StoreCache = {};
+  const tables = getTables(schema);
+  const readonlyStore = getReadonlyStore({
+    encoding,
+    schema,
+    namespaceInfo,
+    db,
+  });
+
+  common.logger.debug({
+    service: "indexing",
+    msg: `Using a ${Math.round(maxSizeBytes / (1024 * 1024))} mB cache.`,
+  });
+
+  /** True if the cache contains the complete state of the store. */
+  let isCacheFull = _isCacheFull;
+
+  /** Number of rows in cache. */
+  let cacheSize = 0;
+  /** Estimated number of bytes used by cache. */
+  let cacheSizeBytes = 0;
+  /** LRU counter. */
+  let totalCacheOps = 0;
+
+  for (const tableName of Object.keys(tables)) {
+    storeCache[tableName] = {
+      insert: {},
+      update: {},
+      find: {},
+    };
+  }
+
+  const getCacheId = (id: UserId, tableName: string): string | number => {
+    if (tables[tableName].table.id[" scalar"] === "hex")
+      return padHex(id as Hex, {
+        size: Math.ceil(((id as Hex).length - 2) / 2),
+        dir: "left",
+      }).toLowerCase();
```

Comment:
What's going on here?

Reply:
All of the cases are equivalent, because of the bytes encoding we are doing in the db: "0xa", "0xA", "0x0a", "0x0A"